Mailer extension

cescoffier opened the pull request:
Implementation of an extension to send emails.

The extension is based on an asynchronous and non-blocking mail client (Vert.x Mail client), so does not use `javax.mail`. 

The extension exposed:
* an imperative API (`Mailer`)
* a reactive API (`ReactiveMailer`)
* the underlying Vert.x Mail Client

It supports:

* sending text emails
* sending attachments
* sending HTML images with inline attachments

The extension contains tests and documentation. 

Of course, it works in native mode too.

cescoffier committed:
Implementation of the mailer extension

cescoffier committed:
Register the mailer extension

cescoffier committed:
Documentation of the mailer extension

cescoffier commented:
@pmlopes @tsegismont can you also have a look?

cescoffier committed:
Use a random port in tests

cescoffier committed:
A bit of cleanup in the tests

* Avoid warnings
* Reduce visibility when possible

tsegismont commented on bom/pom.xml:
```diff
@@ -159,6 +159,7 @@
         <!-- Used for integration tests, to make sure webjars work-->
         <bootstrap.version>3.1.0</bootstrap.version>
         <keycloak.version>6.0.1</keycloak.version>
+        <subethasmtp.version>3.1.7</subethasmtp.version>
```
Is this needed in the BOM since it's only for testing?

cescoffier replied:
Others test dependencies are also there. I've followed the pattern.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+* `io.quarkus.mailer.Mailer` provides the imperative (blocking and synchronous) API;
+* `io.quarkus.mailer.ReactiveMailer` provides the reactive (non-blocking and asynchronous) API
+
+NOTE: `Mailer` is implemented on top of `ReactiveMailer`
```
Perhaps it would be more informative to say there are the same featurewise.

cescoffier replied:
Done.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+@GET
+@Path("/simple")
+public Response sendASimpleEmail() {
+    mailer.send(Mail.withText("to@acme.org", "A simple email from quarkus", "This is my body"));
```
Can you split the snippet in two parts? Mail API usage and JAX-RS method impl?

cescoffier replied:
Done.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+            .thenApply(x -> Response.accepted().build());
+}
+----
+
```
Can you add a statement about how to test the code with `curl` or the browser?

cescoffier replied:
Done.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+----
+
+Attachments can be created from raw bytes (as shown in the snippet) or files.
+When using a file, the content is read asynchronously.
```
This will confuse the blocking API users, I think.

cescoffier replied:
Agreed, let me remove this sentence.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+== Sending HTML emails with inlined attachments
+
+When sending HTML email, you can add inlined attachments.
+For example, you can send an image with your email, and this email will be displayed in the mail content.
```
```suggestion
For example, you can send an image with your email, and this image will be displayed in the mail content.
```

cescoffier replied:
Done.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+
+Note the _content-id_ format and reference.
+When you create the inline attachment, the content-id must be structured as follows: `<id@domain>`.
+When referenced, for instance in the `src` attribute, use `cid:id@domain` (without the `<` and `>`).
```
If the brackets are always added around the `cid`, can we do in user's stead?

cescoffier replied:
I was wondering about this. To be honest, I found this syntax odd, took me a few minutes to understand it. We can check that the given cid starts and ends with `</>` and add them if needed.

cescoffier replied:
Implemented, tested, doc updated.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+@Inject MailClient client;
+----
+
+Are exposed:
```
```suggestion
Three API flavors are exposed:
```

cescoffier replied:
Done.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+* the Axle client (`io.vertx.axle.ext.mail.MailClient`), using `CompletionStage` and Reactive Streams `Publisher`
+* the RX Java 2 client (`io.vertx.reactivex.ext.mail.MailClient`)
+* the bare client (`io.vertx.ext.mail.MailClient`)
+
```
Can you point to the Vert.x extension guide for further details about the different flavors and when they are useful?

cescoffier replied:
Done.

tsegismont commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,205 @@
+
+== Using the underlying Vert.x Mail Client
+
+The Quarkus Mailer is implemented on top of the https://vertx.io/docs/vertx-mail-client/java/[Vert.x Mail Client], providing an asynchronous and non-blocking way to send emails.
```
Can you add a comment why one would want to use the underlying client?

cescoffier replied:
I've added a sentence and a use case (message id retrieval)

tsegismont reviewed (changes requested):
@cescoffier nice work! I've only reviewed the guide, I'll defer to @pmlopes for the implementation

cescoffier committed:
Fix build on JDK 11

For this, I added an exception to banned dependency list, but only in the test scope.

cescoffier committed:
Automatically wrap the content id with < and > if the passed one do not have these

This is to enforce the spec.

cescoffier committed:
Apply the reviews from Thomas (mostly doc)

Also update the section about inline attachment to mention that content id can be automatically wrapped between `<>`.

pmlopes commented:
This is not using vert.x 3.7.1 right? Because it misses `XOAUTH` and it needs reflection config for the the SASL providers? This should be not needed with 3.7.1.

For the rest the code looks fine!

dmlloyd commented on extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java:
```diff
@@ -0,0 +1,60 @@
+        // We must register the auth provider used by the Vert.x mail clients
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, true,
+                "io.vertx.ext.mail.impl.sasl.AuthDigestMD5",
+                "io.vertx.ext.mail.impl.sasl.AuthCramSHA256",
```
There aren't mailers actually using these algorithms are there?  Also we should really unify this with Elytron, at some point.

pmlopes replied:
Actually there are, for example: `Cyrus`, `Dovecot`, `Oracle`, etc... just to name a few. The upstream project `vertx-mail` does not include any security related library so `Elytron` is not a dependency.

dmlloyd replied:
Understood; however, we should only be using one set of SASL implementations if possible.  Is `vertx` using the standard SASL API or are these somehow natively reactive implementations?

cescoffier replied:
@pmlopes ^ can you answer?

gsmet replied:
@cescoffier can you open an issue to deal with this? I think we need the mailer in so I will merge this PR but we need to address @dmlloyd 's comment.

dmlloyd commented on extensions/mailer/runtime/src/main/java/io/quarkus/mailer/Attachment.java:
```diff
@@ -0,0 +1,181 @@
+    private File file;
+    private String description;
+    private String disposition;
+    private byte[] data;
```
Shouldn't we make this stream-oriented somehow?

cescoffier replied:
Done, you can now pass a `Publisher<Byte>`.

dmlloyd commented on extensions/mailer/runtime/src/main/java/io/quarkus/mailer/impl/MailConfig.java:
```diff
@@ -0,0 +1,129 @@
+     * The password.
+     */
+    @ConfigItem
+    public Optional<String> password;
```
We need to do something about clear passwords in config.

FroMage replied:
Like what?

dmlloyd replied:
In WildFly ((cue groaning)) we have a concept of a "credential store" that can be referenced from the configuration.  The credential store would contain all the sensitive stuff and could be protected using various means.  In the Quarkus case we'd probably want to delegate to OpenShift or similar to hold security credentials whenever possible.

In fact we even go a bit farther in WildFly by having a complete separate configuration governing client authentication, which matches outbound requests by protocol, host or IP address, abstract type, etc. and specifies the type of authentication to perform (which might include e.g. OAUTH or Kerberos or other non-user-password strategy) along with all of the associated parameters.  This might or might not be pushing things a bit too far for our use case - but it might be worth discussing anyway.

FroMage replied:
In my experience, if the application can extract the sensitive config by itself, it's only ever just slightly obfuscated, so putting the passwords in plain text does not really hurt.

If outside people have access to the deployment, however, such as by being able to view the source code or browse config files, then the best way to pass those passwords remains like OpenShift does and pass them as env vars from a source that is protected.

So in both cases I'm not sure if we need to do something special to hide passwords. I believe we already support passing config values via env vars.

dmlloyd replied:
Right, the most common problem has less to do with deployment and more to do with e.g. the fact that our issue template asks the user to paste their `application.properties`. ;-)  Many users do not consider configuration files to be sensitive and will happily email them, paste them to the web, etc.  It's better to at least put passwords into a separate file (which can then be `chmod 0400` for example) to help avoid accidents.

FroMage replied:
OK, but in that case let's FINALLY support config files like this:

In `application.properties`
```
quarkus.smtp.host=myhost
# I don't care about the syntax
@include passwords.properties
```

In `passwords.properties`:
```
quarkus.smtp.password=secret
```

Or, you know, make it a convention, and OOTB load `passwords.properties` if it exists.

cescoffier replied:
I've opened https://github.com/quarkusio/quarkus/issues/2648 to discuss it.

dmlloyd commented on extensions/mailer/runtime/src/main/java/io/quarkus/mailer/impl/MailConfig.java:
```diff
@@ -0,0 +1,129 @@
+     * The content of the emails is printed on the console.
+     */
+    @ConfigItem
+    public Optional<Boolean> mock;
```
It's better to just use a primitive `boolean` with a default value than having a tristate like this.

cescoffier replied:
Done.

dmlloyd commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,227 @@
+| password | String | the password
+| ssl | Boolean | Enables or disables SSL
+| trustAll | Boolean | Set whether to trust all certificates
+| maxPoolSize | Integer | Set the maximum pool size
```
Pool of what?

FroMage replied:
Doesn't matter as long as `olympic` is an accepted value

cescoffier replied:
It's the number of connection to the server.

cescoffier replied:
I've rewritten the documentation and updated the javadoc to be clearer.

gsmet reviewed (changes requested):
Added a few minor comments inline.

cescoffier committed:
Add methods to pass a Publisher as data for an attachment

This was suggested by David Lloyd. Note that until the underlying mail client accept ReadStreams, the data is accumulated in memory. However, this change has been made to avoid changing the API later. Upstreams issue is https://github.com/vert-x3/vertx-mail-client/issues/106.

This commit also adds many unit tests around the Mail and Attachment classes.

cescoffier committed:
Avoid using Optional<Boolean> in the config (use bare boolean instead)

Extend the javadoc to indicate the default values.

cescoffier commented on build-parent/pom.xml:
```diff
@@ -849,6 +849,13 @@
+                                            Exception for the in-memory mail server (subethasmtp).
+                                            Only accepted in the test scope
+                                             -->
+                                            <include>javax.activation:activation:*:jar:test</include>
```
That's the dependency I was looking for!

cescoffier commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,227 @@
+| maxPoolSize | Integer | Set the maximum pool size
+| ownHostName | String | The hostname to be used for `HELO/EHLO` and the `Message-ID`
+| keepAlive | Boolean | Set if connection pool is enabled, default is `true`
+| disableEsmtp | Boolean | Set if ESMTP should be tried as first command (EHLO).
```
The comment part was overly complicated, I'm fixing it.

cescoffier committed:
Fix build configuration

* use the recommended javax.activation artifact
* declare the test scope on the quarkus junit5 dependency

cescoffier committed:
Clarify the configuration both in the documentation and javadoc.

* Rework the description of the maximumPoolSize and disableESMTP properties
* Add default values
* Fix some typos
* Add a conclusion indicating that the mailer extension work also in native mode.

cescoffier commented on docs/src/main/asciidoc/sending-emails.adoc:
```diff
@@ -0,0 +1,227 @@
+The retrieved `MailClient` is configured using the configuration key presented above.
+You can also create your own instance, and pass your own configuration.
+
+
```
I've added a conclusion section indicating that the extension works also in native mode.

gsmet reviewed (approved):
Let's have this one in for 0.16.0, it's a great addition.
